_posts/2017-07-05-SpringIO17-Summary.md

```diff
@@ -0,0 +1,773 @@
+
+## Splitting the monolith
+
+The popular route from monolith to a microservice architecture is slicing off bits of functionality.
+This is hard in practice, splitting up a schema is usually the complex part.
+Refactoring out functionality and tables to new services can be hard because of foreign key constraints for example.
+
+### Why we need event-driven microservices
+
+The problem with microservices is that there are no foreign key constraints between services.
+Furthermore, distributed transactions are brittle and distributed systems are hard.
+
+> You will drown in problems you didn't know existed!
+
+Without event-driven microservices and an audit trail you will never know why something went wrong.
+This audit trail allows you to reason about what went wrong and roll back state.
+    
+## Rules for Event-driven microservices
+
+A lot of these rules are from reference applications and Kenny's work with [Chris Richardson](http://microservices.io/).
+
+*Domain events* are a first class citizen, every time you change some piece of data, domain events should be exchanged.
+These events can be used as an audit trail to determine why state changed in a system.
+Each domain event contains a subject with the project aggregate and a payload with immutable data.
+
+{% highlight java %}
+
+@Entity
+@EntityListeners(AuditionEntityListener.class)
+public class ProjectEvent {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long eventId;
+    
+    @Enumerated(EnumType.STRING)
+    private ProjectEventType type;
+    
+    private Project entity;
+    
+    private Map<String, Object> payload;
+
+{% endhighlight %}
+
+To make this way of working accessible to the developers, hypermedia api's need to expose links on the aggregates.
+A traversal list of command and a log of events that happened should be accessible.
+
+**Command handlers** trigger commands on aggregate and then the command is going to generate events.
+Every domain event applies a state transition to an aggregate.
+
+**Event handlers** are going to subscribe to an event and apply changes to an aggregate to change the state.
+In a graph representation, event handlers would be the nodes and events the edges.
+
+**CQRS** is used to create materialized views from streams of events.
+With CQRS you will have a command side and a query side.
+For example the commands might be written to an Apache Kafka event store.
+Then an event processor could be using Spring Cloud Stream to retrieve these events and create a data model.
+The data model is then written to a Data Store like MySQL, where the query side reads the data.
+An API gateway, like Spring Cloud Netflix Zuul, can be put in front so it looks for the consumer like a regular microservice.
+For deploying this application you can combine these components together of scale em independently.
```

Comment:
you can combine these components together of scale em independently =>  or scale them independently